src/grid.py

```diff
@@ -1,40 +1,64 @@
 from numpy import newaxis as na
 
-from .utilities import Vec3, cosd, sind, tand
+from .utilities import Vec3, cosd, sind
 
 
-class Grid(ABC):
-    def __init__(
-        self,
-        turbine_coordinates: List[Vec3],
-        reference_turbine_diameter: float,
-        grid_resolution: int,
-    ) -> None:
-        """
-        Grid should establish domain bounds based on given criteria,
-        and develop three arrays to contain components of the
-        wind velocity.
+NDArrayFloat = npt.NDArray[np.float64]
+NDArrayInt = npt.NDArray[np.int_]
 
-        This could be generalized to any number of dimensions to be
-        used by perhaps a turbulence field.
 
-        The grid will have to be reestablished for each wind direction since the planform
-        area of the farm will be different.
-        """
-        self.turbine_coordinates: List[Vec3] = turbine_coordinates
-        self.reference_turbine_diameter: float = reference_turbine_diameter
-        self.grid_resolution: int = grid_resolution
-        # x are the locations in space in the primary direction (typically the direction of the wind)
-        # y are the locations in space in the lateral direction
-        # z are the locations in space in the vertical direction
-        # u are the velocity components at each point in space
-        # v are the velocity components at each point in space
-        # w are the velocity components at each point in space
-        # all of these arrays are the same size
+@attr.s(auto_attribs=True)
+class Grid(ABC):
+    """
+    Grid should establish domain bounds based on given criteria,
+    and develop three arrays to contain components of the
+    wind velocity.
+
+    This could be generalized to any number of dimensions to be
+    used by perhaps a turbulence field.
+
+    The grid will have to be reestablished for each wind direction since the planform
+    area of the farm will be different.
+
+    Args:
+        turbine_coordinates (`list[Vec3]`): The collection of turbine coordinate (`Vec3`) objects.
+        reference_turbine_diameter (:py:obj:`float`): The reference turbine's rotor diameter.
+        reference_wind_height (:py:obj:`float`): The reference wind height.
+        grid_resolution (:py:obj:`int`): The number of points on each turbine
+    """
+
+    # TODO: We'll want to consider how this expands to multiple turbine types
+    turbine_coordinates: list[Vec3]  # TODO: Should this be an array to match the Farm data?
+    reference_turbine_diameter: float
+    reference_wind_height: float
+    grid_resolution: int
+
+    n_turbines: int = attr.ib(init=False)
+    turbine_coordinates_array: NDArrayFloat = attr.ib(init=False)
+    coordinate_shape: tuple[int, int, int] = attr.ib(init=False)
+    x: NDArrayFloat = attr.ib(init=False)
+    y: NDArrayFloat = attr.ib(init=False)
+    z: NDArrayFloat = attr.ib(init=False)
+
+    def __attrs_post_init__(self) -> None:
+        self.n_turbines = len(self.turbine_coordinates)
```

Comment:
This may be best as a virtual property since it is derived from another property. If `self.turbine_coordinates` changes, we would want to ensure that `n_turbines` also changes.

Reply:
I've used a validator method that enables the value to be changed once `turbine_coordinates` is updated, that way there is no recalculation throughout the code. I also did this for the wind speeds and wind directions. Let me know if this isn't desired.